ciborium/src/simple.rs

```diff
@@ -0,0 +1,402 @@
+//! Contains helper types for dealing with CBOR simple values
+
+use ciborium_ll::{simple, Decoder, Header};
+use serde::{de, forward_to_deserialize_any, ser};
+
+pub(crate) struct SimpleDeserializer<'d, R, D> {
+    parent: core::marker::PhantomData<D>,
+    decoder: &'d mut Decoder<R>,
+}
+
+impl<'d, R, D> SimpleDeserializer<'d, R, D> {
+    pub fn new(decoder: &'d mut Decoder<R>) -> Self {
+        Self {
+            parent: core::marker::PhantomData,
+            decoder,
+        }
+    }
+}
+
+impl<'de, 'd, D: de::Deserializer<'de>, R: ciborium_io::Read> de::Deserializer<'de>
+    for &mut SimpleDeserializer<'d, R, D>
+where
+    R::Error: core::fmt::Debug,
+{
+    type Error = crate::de::Error<R::Error>;
+
+    #[inline]
+    fn deserialize_any<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value, Self::Error> {
+        let offset = self.decoder.offset();
+        let Header::Simple(simple) = self.decoder.pull()? else{
+                return Err(crate::de::Error::semantic(offset, "expected simple"));
+            };
+        visitor.visit_u8(simple)
+    }
+
+    #[inline]
+    fn deserialize_bool<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value, Self::Error> {
+        loop {
+            let offset = self.decoder.offset();
+
+            return match self.decoder.pull()? {
+                Header::Tag(..) => continue,
+                Header::Simple(simple::FALSE) => visitor.visit_bool(false),
+                Header::Simple(simple::TRUE) => visitor.visit_bool(true),
+                _ => Err(crate::de::Error::semantic(offset, "expected bool")),
+            };
+        }
+    }
+
+    forward_to_deserialize_any! {
+        i8 i16 i32 i64 i128
+        u8 u16 u32 u64 u128
+        f32 f64
+        char str string
+        bytes byte_buf
+        seq map
+        struct tuple tuple_struct
+        identifier ignored_any
+        option unit unit_struct newtype_struct enum
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct Error;
+
+impl core::fmt::Display for Error {
+    #[inline]
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl ser::StdError for Error {}
+
+impl ser::Error for Error {
+    fn custom<U: core::fmt::Display>(_msg: U) -> Self {
+        Error
+    }
+}
+
+pub(crate) struct Serializer;
```

Comment:
Why are all the functions for `Serializer` returning an error?

Reply:
The only function returning `Ok` is `serialize_u8`, which is the only legal content of a simple value. It is used in [ciborium/src/ser/mod.rs (`Serializer::serialize_newtype_struct`)](https://github.com/enarx/ciborium/pull/137/commits/19fafbac0d25b42a6f1ef50fcc329da9ca0b5057#diff-bc9bec000d361bc167ed6da072a6726893eea63993e0e4074a8fd879c6389b96R207)